servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/HttpClients.java

```diff
@@ -175,6 +206,65 @@ public static SingleAddressHttpClientBuilder<HostAndPort, InetSocketAddress> for
+     * </ol>
+     * <p>
+     * The returned builder can be customized using {@link SingleAddressHttpClientBuilderProvider}.
+     *
+     * @param host host to connect to, resolved by default using a DNS {@link ServiceDiscoverer} for every new
+     * connection. This will also be used for the {@link HttpHeaderNames#HOST} together with the {@code port}. Use
+     * {@link SingleAddressHttpClientBuilder#unresolvedAddressToHost(Function)} if you want to override that value
+     * or {@link SingleAddressHttpClientBuilder#hostHeaderFallback(boolean)} if you want to disable this behavior.
+     * @param port port to connect to
+     * @return new builder for the address
+     * @see SingleAddressHttpClientBuilderProvider
+     */
+    public static SingleAddressHttpClientBuilder<HostAndPort, InetSocketAddress> forSingleAddressResolveOnDemand(
+            final String host, final int port) {
+        return forSingleAddressResolveOnDemand(HostAndPort.of(host, port));
+    }
+
+    /**
+     * Creates a {@link SingleAddressHttpClientBuilder} for an address with default {@link LoadBalancer} and DNS {@link
+     * ServiceDiscoverer} that will run resolutions before creating a new connection.
+     * <p>
+     * Important side effects to take into account:
+     * <ol>
+     *     <li>The total latency for opening a new connection will be increased by a latency of DNS resolution.</li>
+     *     <li>If the target host has more than one resolved address, created clients loose ability to load balance for
+     *     each request. Instead, the balancing will happen only for every new created connection.</li>
+     *     <li>Created clients won't be able to move/shift traffic based on changes in DNS records until the remote
+     *     server closes existing connections.</li>
+     * </ol>
+     * <p>
+     * The returned builder can be customized using {@link SingleAddressHttpClientBuilderProvider}.
+     *
+     * @param address the {@code UnresolvedAddress} to connect to, resolved using a DNS {@link ServiceDiscoverer} for
+     * every new connection. This address will also be used for the {@link HttpHeaderNames#HOST} header value.
+     * Use {@link SingleAddressHttpClientBuilder#unresolvedAddressToHost(Function)} if you want to override that
+     * value or {@link SingleAddressHttpClientBuilder#hostHeaderFallback(boolean)} if you want to disable this behavior.
+     * @return new builder for the address
+     * @see SingleAddressHttpClientBuilderProvider
+     */
+    public static SingleAddressHttpClientBuilder<HostAndPort, InetSocketAddress> forSingleAddressResolveOnDemand(
+            final HostAndPort address) {
+        return applyProviders(address,
+                new DefaultSingleAddressHttpClientBuilder<>(address, unresolvedServiceDiscoverer()))
+                // Apply after providers to let them see these customizations.
+                .appendConnectionFactoryFilter(ResolvingConnectionFactoryFilter.INSTANCE)
```

Comment:
Heads up: the position of this `ResolvingConnectionFactoryFilter` has some differences. If it's the first one in the chain, all subsequent filters will see already resolved address (like they do today with background resolutions). On the other hand, users of `TransportObserverConnectionFactoryFilter` won't be able to take DNS resolution time into account for opening a new connection. The only way for those users to properly measure time is to apply `TransportObserverConnectionFactoryFilter` via providers.

Reply:
Is this something we should document in the javadoc of the Builder / Factory somewhere?

Reply:
Actually, users will not be able to see it regardless of the order. I documented it as the last item on `DiscoveryStrategy.ON_NEW_CONNECTION` and took some notes to create follow-up tickets.